drivers/input/inport/hardware.c

```diff
@@ -0,0 +1,428 @@
+    #define NEC_PPI_INT_DISABLE  0x09
+    #define NEC_PPI_HC_NO_CLEAR  0x0E
+    #define NEC_PPI_HC_CLEAR     0x0F
+    #define NEC_PPI_DEFAULT_MODE 0x93
+
+#define NEC_BM_INT_RATE      0x4002
+    #define NEC_RATE_120_HZ      0x00
+    #define NEC_RATE_60_HZ       0x01
+    #define NEC_RATE_30_HZ       0x02
+    #define NEC_RATE_15_HZ       0x03
+
+#define NEC_BM_HIRESO_BASE   (PUCHAR)0x61
+
+#define NEC_BUTTON_RIGHT    0x20
+#define NEC_BUTTON_LEFT     0x80
+
+/*
+ * Microsoft InPort
+ */
+#define MS_INPORT_CONTROL    0x00
+    #define INPORT_REG_BTNS    0x00
+    #define INPORT_REG_X       0x01
+    #define INPORT_REG_Y       0x02
+    #define INPORT_REG_MODE    0x07
+    #define INPORT_RESET       0x80
+
+#define MS_INPORT_DATA       0x01
+    #define INPORT_MODE_IRQ    0x01
+    #define INPORT_MODE_BASE   0x10
+    #define INPORT_MODE_HOLD   0x20
+
+#define MS_INPORT_SIGNATURE  0x02
+
+#define MS_BUTTON_MIDDLE   0x01
+#define MS_BUTTON_LEFT     0x02
+#define MS_BUTTON_RIGHT    0x04
+
+/*
+ * Logitech
+ */
+#define LOG_BM_DATA          0x00
+
+#define LOG_BM_SIGNATURE     0x01
+    #define LOG_SIGNATURE_BYTE 0xA5
+
+#define LOG_BM_CONTROL       0x02
+    #define LOG_ENABLE_IRQ     0x00
+    #define LOG_DISABLE_IRQ    0x10
+
+    #define LOG_READ_X_LOW     0x80
+    #define LOG_READ_X_HIGH    0xA0
+    #define LOG_READ_Y_LOW     0xC0
+    #define LOG_READ_Y_HIGH    0xE0
+
+#define LOG_BM_CONFIG        0x03
+    #define LOG_DEFAULT_MODE   0x90
+    #define LOG_CONFIG_BYTE    0x91
+
+#define LOG_BUTTON_RIGHT     0x20
+#define LOG_BUTTON_MIDDLE    0x40
+#define LOG_BUTTON_LEFT      0x80
+
+/* FUNCTIONS ******************************************************************/
+
+VOID
+NTAPI
+InPortDpcForIsr(
+    _In_ PKDPC Dpc,
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _Inout_ PIRP Irp,
+    _In_opt_ PVOID Context)
+{
+    PINPORT_DEVICE_EXTENSION DeviceExtension = DeviceObject->DeviceExtension;
+    KIRQL OldIrql;
+    ULONG DummyInputDataConsumed;
+    INPORT_RAW_DATA RawData;
+    ULONG ButtonDiff;
+
+    UNREFERENCED_PARAMETER(Dpc);
+    UNREFERENCED_PARAMETER(Irp);
+    UNREFERENCED_PARAMETER(Context);
+
+    /* Copy raw data */
+    OldIrql = KeAcquireInterruptSpinLock(DeviceExtension->InterruptObject);
+    RawData = DeviceExtension->RawData;
+    KeReleaseInterruptSpinLock(DeviceExtension->InterruptObject, OldIrql);
+
+    ButtonDiff = DeviceExtension->MouseButtonState ^ RawData.Buttons;
+
+    /* Fill out fields */
+    DeviceExtension->MouseInputData.LastX = RawData.DeltaX;
+    DeviceExtension->MouseInputData.LastY = RawData.DeltaY;
+    DeviceExtension->MouseInputData.ButtonFlags = 0;
+    if (ButtonDiff != 0)
+    {
+        switch (DeviceExtension->MouseType)
+        {
+            case NecBusMouse:
+            {
+                if (ButtonDiff & NEC_BUTTON_LEFT)
+                {
+                    if (RawData.Buttons & NEC_BUTTON_LEFT)
+                        DeviceExtension->MouseInputData.ButtonFlags |= MOUSE_LEFT_BUTTON_UP;
```

Comment:
The NEC and Logitech mouse seem to be up when the flag is set, whereas the MS mouse is down. Is this a bug?
If not, it could make sense to have the flag names reflect that they mean "button is not pressed", or have a comment about this.